src/compiler/checker.ts

```diff
@@ -15217,7 +15217,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const typeArguments = !node ? emptyArray :
                 node.kind === SyntaxKind.TypeReference ? concatenate(type.target.outerTypeParameters, getEffectiveTypeArguments(node, type.target.localTypeParameters!)) :
                 node.kind === SyntaxKind.ArrayType ? [getTypeFromTypeNode(node.elementType)] :
-                map(node.elements, getTypeFromTypeNode);
+                map(node.elements, element => removeMissingType(getTypeFromTypeNode(element), element.kind === SyntaxKind.OptionalType));

```

Comment:
It seems strange to me that this is being done inside of `getTypeArguments`; previously this function effectively just pulled info from the node and that's it, no processing.

Is there no other place this makes sense?

Reply:
I would go even further and say that the optionality should **always** be removed from those type arguments (of optional elements).

When it comes to object properties the optionality is added to the declared type based on the optionality marker on the property, see [here](https://github.com/microsoft/TypeScript/blob/dce7b6d2e7f5390e859ce45ae2bd50f1d68725c9/src/compiler/checker.ts#L10795-L10797). I see a strong parallel between this and the element types. The type argument is like a property's declared type - the optionality is a separate piece of information.

In a somewhat similar manner, in the case of a rest element - the type argument is not an array type! It's an element type of that array and the final array type~ gets created when the type argument meets the element flag.

I think removing this from the type arguments is best because the iterated type of an array comes from the `Symbol.iterator` method. Its signature is defined as `(): IterableIterator<T>` and `T` gets instantiated to the union that contains all type arguments. So at the end, we get something like `(): IterableIterator<string | number | undefined>`. 

We might notice here that the rest type fits into this perfectly because its type is not an array type - so if we just push that to the union we end up with its element type and not an array, its concatenable~ with required elements, and the `Symbol.iterator` method that gets instantiated this way is correct.

It doesn't work well with optional elements in combination with `exactOptionalPropertyTypes` though if the `undefined` is included in the type argument right from the start. After all, at runtime - we won't actually iterate over that `undefined` since `undefined` isn't assignable to an optional element under EOPT. 

So, I run an experiment to remove this at all times and "add back" that optionality at other places (see the commit [here](https://github.com/microsoft/TypeScript/commit/ec69e2bb96c9ee7972820fc4613610d786ecc8f9)). It turned out to be... semi-successful. All the tests pass except the one that I'm trying to fix here.

The problem with this is that the easiest way to add back the optionality is through the created mapper ([here](https://github.com/microsoft/TypeScript/commit/ec69e2bb96c9ee7972820fc4613610d786ecc8f9#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R12865-R12868)). But then the `T -> UnionOfTypeArguments` uses that as well when instantiating the `Symbol.iterator` method. I don't see a clean way to branch this somehow or to remove the missing type afterward. After all, it's not only about what we get internally. The `Symbol.iterator`'s method should get instantiated to, let's say, `(): IterableIterator<string | number>` and not to `(): IterableIterator<string | number | undefined>`. That's publicly~ visible information that can be obtained through types in user programs.

So, for better or worse - I think that it's best to leave things as-is but avoid adding the `missingType` to optional type arguments under EOPT. Perhaps a cleaner way of doing this would be something along those lines:
```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index 512e191cce..874318fb3b 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -15563,7 +15563,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const typeArguments = !node ? emptyArray :
                 node.kind === SyntaxKind.TypeReference ? concatenate(type.target.outerTypeParameters, getEffectiveTypeArguments(node, type.target.localTypeParameters!)) :
                 node.kind === SyntaxKind.ArrayType ? [getTypeFromTypeNode(node.elementType)] :
-                map(node.elements, element => removeMissingType(getTypeFromTypeNode(element), element.kind === SyntaxKind.OptionalType));
+                map(node.elements, element => getTypeFromTypeNode(element.kind === SyntaxKind.OptionalType && exactOptionalPropertyTypes ? (element as OptionalTypeNode).type : element));
             if (popTypeResolution()) {
                 type.resolvedTypeArguments = type.mapper ? instantiateTypes(typeArguments, type.mapper) : typeArguments;
             }
```

I think I like this one better and I am going to push it out in a second.

Reply:
There is also a non-zero chance that this actually makes some of the existing `removeMissingType` calls redundant. I will investigate this.